File: fonts/convert_lucide_icons.py
#!/usr/bin/env python3
"""
Convert Lucide SVG icons to C bitmap arrays for Sharp Memory LCD.
Uses rsvg-convert for SVG rasterization, then ImageMagick for processing.
"""

import subprocess
import sys
import tempfile
import re
from pathlib import Path

# Icon definitions: (svg_file, c_name)
ICONS = [
    ("lucide-battery.svg", "BATTERY_EMPTY"),
    ("lucide-battery-low.svg", "BATTERY_LOW"),
    ("lucide-battery-medium.svg", "BATTERY_MED"),
    ("lucide-battery-full.svg", "BATTERY_FULL"),
    ("lucide-signal-low.svg", "SIGNAL_1"),
    ("lucide-signal-medium.svg", "SIGNAL_2"),
    ("lucide-signal-high.svg", "SIGNAL_3"),
    ("lucide-signal.svg", "SIGNAL_FULL"),
    ("lucide-mail.svg", "MAIL"),
]

SIZE = 24
BYTES_PER_ROW = (SIZE + 7) // 8  # 3 bytes for 24 pixels


def fix_svg_colors(svg_content: str) -> str:
    """Replace currentColor with black for proper rendering."""
    return svg_content.replace('currentColor', '#000000')


def svg_to_bitmap(svg_path: Path, size: int = 24, threshold: int = 128) -> list[list[int]]:
    """Rasterize SVG to monochrome bitmap."""
    # Read and fix the SVG
    svg_content = svg_path.read_text()
    fixed_svg = fix_svg_colors(svg_content)

    with tempfile.NamedTemporaryFile(suffix='.svg', mode='w', delete=False) as f:
        f.write(fixed_svg)
        temp_svg = f.name

    with tempfile.NamedTemporaryFile(suffix='.png', delete=False) as f:
        temp_png = f.name

    try:
        # Render SVG to PNG using rsvg-convert
        subprocess.run([
            'rsvg-convert', '-w', str(size), '-h', str(size),
            temp_svg, '-o', temp_png
        ], check=True, capture_output=True)

        # Use ImageMagick to composite onto white background and get pixel data
        result = subprocess.run([
            'magick', '-size', f'{size}x{size}', 'xc:white',
            temp_png, '-composite', '-colorspace', 'gray', 'txt:-'
        ], capture_output=True, text=True, check=True)

        # Parse pixel data
        bitmap = [[1] * size for _ in range(size)]  # Default white (1)

        for line in result.stdout.strip().split('\n'):
            if line.startswith('#'):
                continue
            # Parse "0,0: (value)  #HEXHEX  gray(N)"
            match = re.match(r'(\d+),(\d+):\s+\((\d+)\)', line)
            if match:
                x, y, val = int(match.group(1)), int(match.group(2)), int(match.group(3))
                if x < size and y < size:
                    # Threshold: below threshold = black (1 = set bit), above = white (0)
                    # We want icon strokes (black in SVG) to have bits SET
                    bitmap[y][x] = 0 if val >= threshold else 1

        return bitmap

    finally:
        Path(temp_svg).unlink(missing_ok=True)
        Path(temp_png).unlink(missing_ok=True)


def bitmap_to_c_array(bitmap: list[list[int]], name: str) -> str:
    """Convert bitmap to C array declaration."""
    lines = []
    lines.append(f"static const uint8_t {name}[{SIZE} * {BYTES_PER_ROW}] = {{")

    for row in bitmap:
        row_bytes = []
        for byte_idx in range(BYTES_PER_ROW):
            byte_val = 0
            for bit in range(8):
                pixel_idx = byte_idx * 8 + bit
                if pixel_idx < SIZE and row[pixel_idx]:
                    byte_val |= (1 << (7 - bit))  # MSB first
            row_bytes.append(f"0x{byte_val:02X}")
        lines.append(f"    {', '.join(row_bytes)},")

    lines.append("};")
    return '\n'.join(lines)


def bitmap_to_ascii(bitmap: list[list[int]]) -> str:
    """Convert bitmap to ASCII art for debugging."""
    result = []
    for row in bitmap:
        line = ''.join('#' if p else '.' for p in row)  # 1 = stroke (#), 0 = background (.)
        result.append(line)
    return '\n'.join(result)


def main():
    script_dir = Path(__file__).parent

    # Check for --debug flag
    debug = '--debug' in sys.argv

    if not debug:
        print("#pragma once")
        print()
        print("// Auto-generated from Lucide SVG icons")
        print("// Run: python3 convert_lucide_icons.py > ../src/icons.h")
        print()
        print("#include <stdint.h>")
        print()
        print("namespace Icons {")
        print()
        print(f"constexpr int SIZE = {SIZE};")
        print()

    # Add empty signal icon (no bars) - all zeros = no strokes
    if not debug:
        print("// Empty signal (no bars)")
        empty_bitmap = [[0] * SIZE for _ in range(SIZE)]
        print(bitmap_to_c_array(empty_bitmap, "SIGNAL_0"))
        print()

    for svg_file, c_name in ICONS:
        svg_path = script_dir / svg_file
        if not svg_path.exists():
            print(f"// WARNING: {svg_file} not found", file=sys.stderr)
            continue

        bitmap = svg_to_bitmap(svg_path, SIZE)

        if debug:
            print(f"=== {c_name} ({svg_file}) ===")
            print(bitmap_to_ascii(bitmap))
            print()
        else:
            print(f"// {svg_file}")
            print(bitmap_to_c_array(bitmap, c_name))
            print()

    if not debug:
        print("}  // namespace Icons")


if __name__ == "__main__":
    main()

File: firmware/artifacts/fonts/convert_lucide_icons.mjs
// Convert Lucide SVG icons to the 24x24 1-bit C arrays in src/icons.h.
//
// Same output as convert_lucide_icons.py, but rasterised with resvg (wasm) so
// it needs neither rsvg-convert nor ImageMagick — only `npm install` here.
//
//   cd artifacts/fonts && npm install && node convert_lucide_icons.mjs > ../../src/icons.h
//   node convert_lucide_icons.mjs --debug     # ASCII art to stderr instead
//   node convert_lucide_icons.mjs --all       # re-rasterise icons already in icons.h too
//
// By default icons that already exist in src/icons.h are copied through
// byte-for-byte and only new names are rasterised. resvg and rsvg/ImageMagick
// anti-alias differently (a few dozen pixels per icon after thresholding), and
// the existing icons were tuned by eye on the panel, so a regenerate should not
// quietly restyle them. --all opts into that.
//
// Bit set = stroke (drawn in black on the panel's white, or white on a black
// button; the renderer decides). MSB first, 3 bytes per row.

import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { createRequire } from "node:module";
import { Resvg, initWasm } from "@resvg/resvg-wasm";

const require = createRequire(import.meta.url);
const here = dirname(fileURLToPath(import.meta.url));

// (svg file, C name) — order is the order in icons.h
const ICONS = [
  ["lucide-battery.svg", "BATTERY_EMPTY"],
  ["lucide-battery-low.svg", "BATTERY_LOW"],
  ["lucide-battery-medium.svg", "BATTERY_MED"],
  ["lucide-battery-full.svg", "BATTERY_FULL"],
  ["lucide-signal-low.svg", "SIGNAL_1"],
  ["lucide-signal-medium.svg", "SIGNAL_2"],
  ["lucide-signal-high.svg", "SIGNAL_3"],
  ["lucide-signal.svg", "SIGNAL_FULL"],
  ["lucide-mail.svg", "MAIL"],
  ["lucide-send.svg", "SEND"],
  ["lucide-trash-2.svg", "TRASH"],
  ["lucide-images.svg", "IMAGES"],
  ["lucide-arrow-right.svg", "ARROW_RIGHT"],
  ["lucide-camera.svg", "CAMERA"],
];

const SIZE = 24;
const BYTES_PER_ROW = 3;
const THRESHOLD = 128;

await initWasm(readFileSync(require.resolve("@resvg/resvg-wasm/index_bg.wasm")));

function rasterise(file) {
  const svg = readFileSync(join(here, file), "utf8").replaceAll("currentColor", "#000000");
  const png = new Resvg(svg, {
    fitTo: { mode: "width", value: SIZE },
    background: "#ffffff",
  }).render();
  const { width, height } = png;
  const rgba = png.pixels; // straight RGBA, white already composited
  const bitmap = [];
  for (let y = 0; y < SIZE; y++) {
    const row = [];
    for (let x = 0; x < SIZE; x++) {
      let v = 255;
      if (x < width && y < height) {
        const i = (y * width + x) * 4;
        v = (rgba[i] + rgba[i + 1] + rgba[i + 2]) / 3;
      }
      row.push(v < THRESHOLD ? 1 : 0);
    }
    bitmap.push(row);
  }
  return bitmap;
}

function toCArray(bitmap, name) {
  const lines = [`static const uint8_t ${name}[${SIZE} * ${BYTES_PER_ROW}] = {`];
  for (const row of bitmap) {
    const bytes = [];
    for (let b = 0; b < BYTES_PER_ROW; b++) {
      let v = 0;
      for (let bit = 0; bit < 8; bit++) {
        const px = b * 8 + bit;
        if (px < SIZE && row[px]) v |= 1 << (7 - bit);
      }
      bytes.push("0x" + v.toString(16).toUpperCase().padStart(2, "0"));
    }
    lines.push(`    ${bytes.join(", ")},`);
  }
  lines.push("};");
  return lines.join("\n");
}

const debug = process.argv.includes("--debug");
const all = process.argv.includes("--all");

// Existing arrays from the committed header, keyed by C name.
const existing = new Map();
try {
  const current = readFileSync(join(here, "../../src/icons.h"), "utf8");
  for (const m of current.matchAll(/static const uint8_t (\w+)\[24 \* 3\] = \{[\s\S]*?\};/g)) {
    existing.set(m[1], m[0]);
  }
} catch {
  // No header yet: everything gets rasterised.
}

const out = [];
if (!debug) {
  out.push("#pragma once", "",
    "// Auto-generated from Lucide SVG icons",
    "// Run from artifacts/fonts: npm install && node convert_lucide_icons.mjs > ../../src/icons.h",
    "", "#include <stdint.h>", "", "namespace Icons {", "", `constexpr int SIZE = ${SIZE};`, "",
    "// Empty signal (no bars)",
    toCArray(Array.from({ length: SIZE }, () => new Array(SIZE).fill(0)), "SIGNAL_0"), "");
}
for (const [file, name] of ICONS) {
  if (!debug && !all && existing.has(name)) {
    out.push(`// ${file}`, existing.get(name), "");
    continue;
  }
  const bitmap = rasterise(file);
  if (debug) {
    process.stderr.write(`${name} (${file})\n` + bitmap.map(r => r.map(p => (p ? "#" : ".")).join("")).join("\n") + "\n\n");
  } else {
    out.push(`// ${file}`, toCArray(bitmap, name), "");
  }
}
if (!debug) {
  out.push("}  // namespace Icons", "");
  process.stdout.write(out.join("\n"));
}
